src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -2134,23 +2144,80 @@ void EmitAnchors(RegexNode node)
             // Emits the code to handle a multiple-character match.
             void EmitMultiChar(RegexNode node)
             {
-                // if (textSpanPos + node.Str.Length >= textSpan.Length) goto doneLabel;
-                // if (node.Str[0] != textSpan[textSpanPos]) goto doneLabel;
-                // if (node.Str[1] != textSpan[textSpanPos+1]) goto doneLabel;
-                // ...
-                EmitSpanLengthCheck(node.Str!.Length);
-                for (int i = 0; i < node.Str!.Length; i++)
+                ReadOnlySpan<char> s = node.Str;
+
+                // Emit the length check for the whole string.  If the generated code gets past this point,
+                // we know the span is at least textSpanPos + s.Length long.
+                EmitSpanLengthCheck(s.Length);
+
+                // If we're doing a case-insensitive comparison, we need to lower case each character,
+                // so we just go character-by-character.  But if we're not, we try to process multiple
+                // characters at a time; this is helpful not only for throughput but also in reducing
+                // the amount of IL and asm that results from this unrolling.
+                if (!IsCaseInsensitive(node))
+                {
+                    void EmitTextSpanOffsetAsByte()
+                    {
+                        // Unsafe.As<char, byte>(ref Unsafe.Add(ref MemoryMarshal.GetReference(textSpan), textSpanPos))
+                        Ldloc(textSpanLocal);
+                        Call(s_memoryMarshalGetReference);
+                        if (textSpanPos > 0)
+                        {
+                            Ldc(textSpanPos);
+                            Call(s_unsafeAddChar);
+                        }
+                        Call(s_unsafeCharAsByte);
```

Comment:
This is unnecessary. `Unsafe.As` is no-op.

Reply:
Nice. Thanks.